featurebyte/query_graph/sql/tile_compute.py

```diff
@@ -234,124 +257,98 @@ def epoch_seconds_to_timestamp(num_seconds: int) -> pd.Timestamp:
-    frequency : int
-        Frequency in feature job setting
-    time_modulo_frequency : int
-        Time modulo frequency in feature job setting
-    blind_spot : int
-        Blind spot in feature job setting
+    adapter : BaseAdapter
+        Instance of BaseAdapter for generating engine specific SQL
+    tile_info : TileGenSql
+        Tile table information
+    request_table_name : str
+        Name of the request table
     window : Optional[int]
-        Window size. If None, it means the feature requires unbounded window.
+        Window size in seconds. None for features with an unbounded window.
 
     Returns
     -------
     Select
     """
+
+    def get_tile_boundary(point_in_time_expr: Expression) -> Expression:
+        previous_job_epoch_expr = get_previous_job_epoch_expr(
+            adapter.to_epoch_seconds(point_in_time_expr), tile_info
+        )
+        return expressions.Anonymous(
+            this="TO_TIMESTAMP",
+            expressions=[expressions.Sub(this=previous_job_epoch_expr, expression=blind_spot)],
+        )
+
+    blind_spot = make_literal_value(tile_info.blind_spot)
+    time_modulo_frequency = make_literal_value(tile_info.time_modulo_frequency)
+
     if window:
-        num_tiles = int(window // frequency)
+        num_tiles = int(window // tile_info.frequency)
     else:
         num_tiles = None
-    start_date, end_date = compute_start_end_date_from_point_in_time(
-        point_in_time_list,
-        frequency=frequency,
-        time_modulo_frequency=time_modulo_frequency,
-        blind_spot=blind_spot,
-        num_tiles=num_tiles,
+
+    # Tile end date is determined from the latest point in time per entity
+    end_date_expr = get_tile_boundary(
+        expressions.Max(this=expressions.Identifier(this=SpecialColumnName.POINT_IN_TIME.value))
     )
-    out_expr = sql_template.render(
-        {
-            InternalName.TILE_START_DATE_SQL_PLACEHOLDER: str(start_date),
-            InternalName.TILE_END_DATE_SQL_PLACEHOLDER: str(end_date),
-        },
-        as_str=False,
+
+    if num_tiles:
+        minus_num_tiles_in_microseconds = expressions.Mul(
+            this=TimedeltaExtractNode.convert_timedelta_unit(
+                expressions.Mul(
+                    this=make_literal_value(num_tiles),
+                    expression=make_literal_value(tile_info.frequency),
+                ),
+                "second",
+                "microsecond",
+            ),
+            expression=make_literal_value(-1),
+        )
+        # Tile start date is determined from the earliest point in time per entity minus the largest
+        # feature window
+        start_date_expr = adapter.dateadd_microsecond(
+            minus_num_tiles_in_microseconds,
+            get_tile_boundary(
+                expressions.Min(
+                    this=expressions.Identifier(this=SpecialColumnName.POINT_IN_TIME.value)
+                )
+            ),
+        )
```

Comment:
Does it make sense to move the logic here to the `get_earliest_tile_start_date_expr` also?

Reply:
The function `get_earliest_tile_start_date_expr` is also used by `tile_cache.py`, but this part of subtracting number of tiles is specific to feature preview only.